gcc/rust/expand/rust-macro-builtins-helpers.cc

```diff
@@ -0,0 +1,256 @@
+	 && parser.peek_current_token ()->get_id () != END_OF_FILE)
+    {
+      auto expr = parser.parse_expr (AST::AttrVec (), restrictions);
+      // something must be so wrong that the expression could not be parsed
+      rust_assert (expr);
+      result.push_back (std::move (expr));
+
+      auto next_token = parser.peek_current_token ();
+      if (!parser.skip_token (COMMA) && next_token->get_id () != last_token_id)
+	{
+	  rust_error_at (next_token->get_locus (), "expected token: %<,%>");
+	  // TODO: is this recoverable? to avoid crashing the parser in the next
+	  // fragment we have to exit early here
+	  has_error = true;
+	  return empty_expr;
+	}
+    }
+
+  return result;
+}
+
+std::unique_ptr<AST::Expr>
+parse_single_string_literal (BuiltinMacro kind,
+			     AST::DelimTokenTree &invoc_token_tree,
+			     location_t invoc_locus, MacroExpander *expander)
+{
+  MacroInvocLexer lex (invoc_token_tree.to_token_stream ());
+  Parser<MacroInvocLexer> parser (lex);
+
+  auto last_token_id = macro_end_token (invoc_token_tree, parser);
+
+  std::unique_ptr<AST::LiteralExpr> lit_expr = nullptr;
+  std::unique_ptr<AST::MacroInvocation> macro_invoc = nullptr;
+
+  if (parser.peek_current_token ()->get_id () == STRING_LITERAL)
+    {
+      lit_expr = parser.parse_literal_expr ();
+      parser.maybe_skip_token (COMMA);
+      if (parser.peek_current_token ()->get_id () != last_token_id)
+	{
+	  lit_expr = nullptr;
+	  rust_error_at (invoc_locus, "macro takes 1 argument");
+	}
+    }
+  else if (parser.peek_current_token ()->get_id () == last_token_id)
+    rust_error_at (invoc_locus, "macro takes 1 argument");
+  else
+    {
+      macro_invoc = parser.parse_macro_invocation (AST::AttrVec ());
+
+      parser.maybe_skip_token (COMMA);
+      if (parser.peek_current_token ()->get_id () != last_token_id)
+	{
+	  lit_expr = nullptr;
+	  rust_error_at (invoc_locus, "macro takes 1 argument");
+	}
+
+      if (macro_invoc != nullptr)
+	{
+	  auto path_str = make_macro_path_str (kind);
+
+	  auto pending_invocations
+	    = std::vector<std::unique_ptr<AST::MacroInvocation>> ();
+
+	  pending_invocations.push_back (std::move (macro_invoc));
+
+	  return AST::MacroInvocation::Builtin (
+	    kind,
+	    AST::MacroInvocData (AST::SimplePath ({AST::SimplePathSegment (
+				   path_str, invoc_locus)}),
+				 std::move (invoc_token_tree)),
+	    {}, invoc_locus, std::move (pending_invocations));
+	}
+      else
+	{
+	  rust_error_at (invoc_locus, "argument must be a string literal or a "
+				      "macro which expands to a string");
+	}
+    }
+
+  parser.skip_token (last_token_id);
+
+  return std::unique_ptr<AST::Expr> (std::move (lit_expr));
+}
+
+std::string
+source_relative_path (std::string path, location_t locus)
+{
+  std::string compile_fname = LOCATION_FILE (locus);
+
+  auto dir_separator_pos = compile_fname.rfind (file_separator);
+
+  /* If there is no file_separator in the path, use current dir ('.').  */
```

Comment:
nit: you're mixing both C and C++ comment style. I don't think we have a strict rule, but I think it's best to be a bit uniform in the same file, even more when everything is new :)

Reply:
I'll resolve to the C++ // style

Reply:
I did a bit of grepping in the gcc/rust/expand folder and saw that there is a lot of C style commenting

```
jjasmine@jjasmine ~/C/g/g/r/expand (split_up_rust_macro_builtsin)> grep -F "/*" * -l
rust-cfg-strip.cc
rust-cfg-strip.h
rust-derive-clone.cc
rust-derive-clone.h
rust-derive-copy.cc
rust-derive-copy.h
rust-derive.h
rust-expand-visitor.cc
rust-expand-visitor.h
rust-macro-builtins-helpers.cc
rust-macro-builtins-helpers.h
rust-macro-builtins-trait.cc
rust-macro-builtins-utility.cc
rust-macro-builtins.cc
rust-macro-builtins.h
rust-macro-expand.cc
rust-macro-expand.h
rust-macro-substitute-ctx.h
rust-proc-macro-invoc-lexer.h
rust-proc-macro.cc
rust-proc-macro.h
```
I'm not sure what i should do in the case of these.